### DISCUSSION TITLE
Running predictions against a different backend to the model build

### DISCUSSION THREAD

**Discussion by richard-a-lott** [moj-analytical-services/splink#1424] [Q&A]
Hi,

I've built a model using Spark (Glue) and am now attempting to run the prediction, however I wanted to do that via Athena, as it can be done in a more interactive way. The prediction failed due to the model dialect being spark, and athena being presto based.

Is it possible to switch the model to a different dialect, or is there another way to achieve what I want?

**Comment by RossKen** [comment]
Hey @richard-a-lott this is a feature we have on the backlog #1281 but we haven't got around to it yet. However, knowing that it is useful for other users is helpful for our prioritisation.

For now (assuming you are using comparison levels that exist across both backends), I reckon manually copying across the m and u values from your spark settings JSON to an Athena one.

**Reply by richard-a-lott** [reply]
Ok good to know, thanks @RossKen !

**Reply by RobinL** [reply]
Haven't tested but something roughly like this should do the trick.  Providing a duckdb->athena because i had an easier example to hand, but spark->athena (modify the `translate()` fn) should also work.

Of course, all the functions you're using in Spark need to be available in Athena.  So for example, if you use `jaro_winkler` in Spark, there's no translation to Athena because the `jaro_winkler` function simply doesn't exist in Athena.

```python
# from try_settings import settings_dict
import pandas as pd
import json
import sqlglot

from splink.duckdb.comparison_library import (
    exact_match,
    levenshtein_at_thresholds,
)
from splink.duckdb.linker import DuckDBLinker

df = pd.read_csv("./tests/datasets/fake_1000_from_splink_demos.csv")


settings = {
    "probability_two_random_records_match": 0.01,
    "link_type": "dedupe_only",
    "blocking_rules_to_generate_predictions": [
        "l.first_name = r.first_name",
        "l.surname = r.surname",
    ],
    "comparisons": [
        levenshtein_at_thresholds("first_name", 2),
        exact_match("surname"),
        exact_match("dob"),
        exact_match("city", term_frequency_adjustments=True),
        exact_match("email"),
    ],
    "retain_matching_columns": False,
    "retain_intermediate_calculation_columns": False,
    "additional_columns_to_retain": ["group"],
    "max_iterations": 10,
    "em_convergence": 0.01,
}


linker = DuckDBLinker(df, settings, connection=":memory:")


# train model

linker.save_model_to_json("settings.json", overwrite=True)

with open("settings.json", "r") as f:
    settings_from_json = json.load(f)


brs = settings_from_json["blocking_rules_to_generate_predictions"]

settings_from_json['sql_dialect'] = "athena"

def translate(sql):
    return sqlglot.parse_one(sql, read="duckdb").sql(dialect="presto")

brs = [translate(br) for br in brs]
settings_from_json["blocking_rules_to_generate_predictions"]  = brs



for comparison in settings_from_json['comparisons']:
    # Iterate over 'comparison_levels' list in each 'comparison' dictionary
    for comparison_level in comparison['comparison_levels']:
        try:
            comparison_level['sql_condition'] = translate(comparison_level['sql_condition'])    
        except:
            print(f"failed to translate {comparison_level['sql_condition']}")


settings_from_json
```

Also note that for truly interactive examples on smaller numbers of records, you may find the [def find_matches_to_new_records](https://github.com/moj-analytical-services/splink/blob/b1ff5591161039547b1df586a80253971957b2bd/splink/linker.py#L1710-L1711) function useful, which you could maybe use in duckdb.   See [here](https://moj-analytical-services.github.io/splink/demos/examples/duckdb/real_time_record_linkage.html)

**Reply by richard-a-lott** [reply]
That's exactly what I needed. I've never heard of sqlglot before. It should have realised someone would have written a translator already! Thanks!

> the jaro_winkler function simply doesn't exist in Athena

Ah I'd not realised. I'll definitely try out the 'find_matches_to_new_records' method, I'm not sure if duckdb will handle my 11M data set though.

**Reply by RobinL** [reply]
Duckdb would definitely struggle with the model training and prediction for 11m, but for finding matches to a single (or handful) of records using 'find_matches_to_new_records' it should be totally fine

**Reply by richard-a-lott** [reply]
Cool thanks. I'll give that a try.

**Reply by richard-a-lott** [reply]
Hi @RobinL, I'm afraid this conversion isn't actually working. I've tried using a variety of sqlglot methods, but it seems to be failing to convert functions such as months_between, to_timestamp and jaro_winkler from spark to duckdb. I've had to go ahead and do it manually.

I know this isn't your problem, I'll consider raising it with sqlglot, I just wanted to note it here for reference.

**Reply by RossKen** [reply]
Ah, I think that will be a Splink issue rather than sqlglot as those functions are defined in Splink/as UDFs and differ across backends. So to do a conversion will probably take more than just the sqlglot functionality

**Reply by richard-a-lott** [reply]
Ah ok that makes sense! However there are a couple of spark functions that sqlglot wasn't recognising (months_between and to_timestamp) so I have raised that as an issue against them:
https://github.com/tobymao/sqlglot/issues/1928
